src/pudl/etl/check_foreign_keys.py

```diff
@@ -50,15 +61,143 @@ def pudl_check_fks(logfile: pathlib.Path, loglevel: str):
+    ):
+        """Initialize a new ForeignKeyError object."""
+        self.child_table = child_table
+        self.parent_table = parent_table
+        self.foreign_key = foreign_key
+        self.rowids = rowids
+
+    def __str__(self):
+        """Create string representation of ForeignKeyError object."""
+        return (
+            f"Foreign key error for table: {self.child_table} -- {self.parent_table} "
+            f"{self.foreign_key} -- on rows {self.rowids}\n"
+        )
+
+    def __eq__(self, other):
+        """Compare a ForeignKeyError with another object."""
+        if isinstance(other, ForeignKeyError):
+            return (
+                (self.child_table == other.child_table)
+                and (self.parent_table == other.parent_table)
+                and (self.foreign_key == other.foreign_key)
+                and (self.rowids == other.rowids)
+            )
+        return False
+
+
+class ForeignKeyErrors(sa.exc.SQLAlchemyError):  # noqa: N818
+    """Raised when data in a database violate multiple foreign key constraints."""
+
+    def __init__(self, fk_errors: list[ForeignKeyError]):
+        """Initialize a new ForeignKeyErrors object."""
+        self.fk_errors = fk_errors
+
+    def __str__(self):
+        """Create string representation of ForeignKeyErrors object."""
+        fk_errors = [str(x) for x in self.fk_errors]
+        return "\n".join(fk_errors)
+
+    def __iter__(self):
+        """Iterate over the fk errors."""
+        return self.fk_errors
+
+    def __getitem__(self, idx):
+        """Index the fk errors."""
+        return self.fk_errors[idx]
+
+
+def _get_fk_list(engine: sa.Engine, table: str) -> pd.DataFrame:
+    """Retrieve a dataframe of foreign keys for a table.
+
+    Description from the SQLite Docs: 'This pragma returns one row for each foreign
+    key constraint created by a REFERENCES clause in the CREATE TABLE statement of
+    table "table-name".'
+
+    The PRAGMA returns one row for each field in a foreign key constraint. This
+    method collapses foreign keys with multiple fields into one record for
+    readability.
+    """
+    with engine.begin() as con:
+        table_fks = pd.read_sql_query(f"PRAGMA foreign_key_list({table});", con)
+
+    # Foreign keys with multiple fields are reported in separate records.
+    # Combine the multiple fields into one string for readability.
+    # Drop duplicates so we have one FK for each table and foreign key id
+    table_fks["fk"] = table_fks.groupby("table")["to"].transform(
+        lambda field: "(" + ", ".join(field) + ")"
+    )
+    table_fks = table_fks[["id", "table", "fk"]].drop_duplicates()
+
+    # Rename the fields so we can easily merge with the foreign key errors.
+    table_fks = table_fks.rename(columns={"id": "fkid", "table": "parent"})
+    table_fks["table"] = table
+    return table_fks
+
+
+def check_foreign_keys(engine: sa.Engine):
```

Comment:
Pulled the foreign key check logic out of the io-manager and put it here directly, which seemed as good a place as anywhere to put it.

Reply:
Yeah, that was my original idea as well. I do not like the name clash in that you end up with `pudl.etl.check_foreign_keys.check_foreign_keys()`.

In the end, I moved this to `helpers.py`. That said, there doesn't really seem to be a good place in the codebase for miscellaneous stuff and that might be useful, e.g. `pudl.misc` where you can have assortment of small modules that don't naturally fit elsewhere. Putting this into `check_foreign_keys.py` is a bit awkward IMO, because of the naming as well as the fact that this module is CLI and I think that mixing CLI/binary capabilities and library capabilities (something you'd be importing elsewhere) seems a little dirty.

This is just a thought and I would not block this PR for this, as this can be easily refactored in the future.

Reply:
This is fine for now. I think I added `check_foreign_keys` to the `SQLiteIOManager` class because I imaged we might have different methods for checking referential integrity for each IO Manager. For example, if we decided to drop sqlite and rely on a combination of parquet files and duckdb we'd need to create a new method of enforcing referential integrity.

Reply:
There's also some actual data transformation creeping into the `etl` subpackage which I thought was just supposed to be laying out the Dagster assets / config stuff (see `glue_assets.py`)